src/utils.ts

```diff
@@ -45,45 +58,41 @@ export function spawnProcess(command: string, args: string[], options: childProc
     });
     child.on('close', (exitCode) => {
       if (exitCode !== 0) {
-        reject(new SpawnError(`${command} ${join(' ', args)} failed with code ${exitCode}`, stdout, stderr));
+        reject(new SpawnError(`${command} ${args.join(' ')} failed with code ${exitCode}`, stdout, stderr));
       } else {
         resolve({ stdout, stderr });
       }
     });
   });
 }
 
+const rootOf = (p: string) => path.parse(path.resolve(p)).root;
+const isPathRoot = (p: string) => rootOf(p) === path.resolve(p);
+const findUpIO = (name: string, directory = process.cwd()): IOO.IOOption<string> =>
+  pipe(path.resolve(directory), (dir) =>
+    pipe(
+      safeFileExistsIO(path.join(dir, name)),
+      IO.chain((exists: boolean) =>
+        exists ? IOO.some(dir) : isPathRoot(dir) ? IOO.none : findUpIO(name, path.dirname(dir))
+      )
+    )
+  );
+
 /**
  * Find a file by walking up parent directories
  */
-export function findUp(names: string | string[], directory: string = process.cwd()): string | undefined {
-  const absoluteDirectory = path.resolve(directory);
-
-  if (typeof names === 'string') {
-    names = [names];
-  }
-
-  /* For vs. .forEach so it can exit when we get a hit. */
-  for (const name of names) {
-    if (fs.existsSync(path.join(directory, name))) {
-      return directory;
-    }
-  }
-
-  const { root } = path.parse(absoluteDirectory);
-  if (absoluteDirectory === root) {
-    return undefined;
-  }
-
-  return findUp(names, path.dirname(absoluteDirectory));
-}
+export const findUp = (name: string) => pipe(findUpIO(name), IOO.toUndefined)();
 
 /**
  * Forwards `rootDir` or finds project root folder.
  */
-export function findProjectRoot(rootDir?: string): string | undefined {
-  return rootDir ?? findUp(['yarn.lock', 'package-lock.json']);
-}
+export const findProjectRoot = (rootDir?: string) =>
+  pipe(
+    IOO.fromNullable(rootDir),
+    IOO.fold(() => findUpIO('yarn.lock'), IOO.of),
+    IOO.fold(() => findUpIO('package-lock.json'), IOO.of),
```

Comment:
Should `pnpm-lock.yaml` be here too? I know support for pnpm isn't great at the moment.

Reply:
seems yes, thanks.